File: JavaScript/Endpoint Examples/Multipart Payload/zugferd-pdf.js
// Create a ZUGFeRD / Factur-X PDF/A-3 invoice from XML and an existing invoice PDF.
// pdfRest preserves the supplied PDF when it agrees with the canonical XML.
const axios = require("axios");
const FormData = require("form-data");
const fs = require("fs");

const apiUrl = "https://api.pdfrest.com";
// const apiUrl = "https://eu-api.pdfrest.com"; // EU/GDPR service
const apiKey = "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx";
const invoiceXml = "/path/to/invoice.xml";
const invoicePdf = "/path/to/invoice.pdf";

const form = new FormData();
form.append("file", fs.createReadStream(invoiceXml));
form.append("pdf_file", fs.createReadStream(invoicePdf));
  // Fallback: generate a replacement PDF when the supplied PDF is mismatched or cannot be fully confirmed.
  form.append("regenerate_pdf", "true");
  // These styles apply only to that fallback-generated PDF; they do not alter a preserved PDF.
form.append("render_options", JSON.stringify({
  locale: "de-DE", label_language: "de", font: "arial", bold_font: "arialbold",
  accent_color_rgb: [0, 92, 171],
}));
form.append("output", "zugferd_invoice");

axios.post(`${apiUrl}/zugferd-pdf`, form, {
  headers: { "Api-Key": apiKey, ...form.getHeaders() }, maxBodyLength: Infinity,
}).then((response) => console.log(JSON.stringify(response.data, null, 2)))
  .catch((error) => { console.error(error.response?.data || error.message); process.exitCode = 1; });

File: Java/Endpoint Examples/Multipart Payload/ZugferdPdf.java
import io.github.cdimascio.dotenv.Dotenv;
import java.io.File;
import java.io.IOException;
import okhttp3.*;
import org.json.JSONObject;

// Create a ZUGFeRD / Factur-X PDF/A-3 invoice from invoice XML.
public class ZugferdPdf {
  private static final String API_URL = "https://api.pdfrest.com";
  private static final String DEFAULT_API_KEY = "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx";

  public static void main(String[] args) throws IOException {
    File invoiceXml = new File(args.length > 0 ? args[0] : "/path/to/invoice.xml");
    String apiKey = Dotenv.configure().ignoreIfMalformed().ignoreIfMissing().load().get("PDFREST_API_KEY", DEFAULT_API_KEY);
    JSONObject options = new JSONObject().put("locale", "de-DE").put("label_language", "de").put("font", "arial").put("bold_font", "arialbold").put("accent_color_rgb", new int[] {0, 92, 171});
    MultipartBody body = new MultipartBody.Builder().setType(MultipartBody.FORM)
        .addFormDataPart("file", invoiceXml.getName(), RequestBody.create(invoiceXml, MediaType.parse("application/xml")))
        .addFormDataPart("render_options", options.toString()).addFormDataPart("output", "zugferd_invoice").build();
    send(new Request.Builder().url(API_URL + "/zugferd-pdf").header("Api-Key", apiKey).post(body).build());
  }

  private static void send(Request request) throws IOException {
    try (Response response = new OkHttpClient().newCall(request).execute()) {
      String body = response.body() == null ? "" : response.body().string(); System.out.println(body);
      if (!response.isSuccessful()) System.exit(1);
    }
  }
}
